### PR TITLE
initial end-to-end tests for streaming, reporting, generating

### PR DESCRIPTION
**Pull request by zenmoto**
Start on end-to-end tests for search commands

### CODE REVIEW

File: tests/searchcommands/chunked_data_stream.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        header = self.stream.readline()
+        if len(header) == 0:
+            raise EOFError
+
+        while len(header) > 0 and header.strip() == b'':
+            header = self.stream.readline()  # Skip empty lines
+        version, meta, data = header.rstrip().split(b',')
+        metabytes = self.stream.read(int(meta))
+        databytes = self.stream.read(int(data))
+        print("sent")
```
Comment (amysutedja) [inline tests/searchcommands/chunked_data_stream.py:50]: Should probably be removed or go to `stderr`.

File: tests/searchcommands/chunked_data_stream.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        return ChunkedDataStreamIter(self)
+
+    def __init__(self, stream):
+        empty = stream.read(0)
+        assert isinstance(empty, bytes)
+        self.stream = stream
+
+    def read_chunk(self):
+        header = self.stream.readline()
+        if len(header) == 0:
```
Comment (amysutedja) [inline tests/searchcommands/chunked_data_stream.py:42]: Should this be moved after the "skip empty lines" logic?

File: tests/searchcommands/chunked_data_stream.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        version, meta, data = header.rstrip().split(b',')
+        metabytes = self.stream.read(int(meta))
+        databytes = self.stream.read(int(data))
+        print("sent")
+        return Chunk(version, metabytes, databytes)
+
+
+def build_chunk(kv, data=None):
+    metadata = six.ensure_binary(json.dumps(kv), 'utf-8')
+    data_output = _build_data_csv(data)
```
Comment (amysutedja) [inline tests/searchcommands/chunked_data_stream.py:56]: IMO we should also have `six.ensure_binary` here as well.

File: tests/searchcommands/chunked_data_stream.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    return b"chunked 1.0,%d,%d\n%s%s" % (len(metadata), len(data_output), metadata, data_output)
+
+
+def build_empty_searchinfo():
+    return {
```
Comment (amysutedja) [inline tests/searchcommands/chunked_data_stream.py:61]: Not a constant?

File: tests/searchcommands/chunked_data_stream.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    return build_chunk({'action': 'execute', 'finished': finished}, data)
+
+
+def _build_data_csv(data):
+    if data is None:
+        return b''
+    if isinstance(data, bytes):
+        return data
+    if six.PY2:
+        csvout = io.BytesIO()
```
Comment (amysutedja) [inline tests/searchcommands/chunked_data_stream.py:86]: I think `six` has an alias that makes `BytesIO` do the right thing here.

File: tests/searchcommands/chunked_data_stream.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    else:
+        csvout = io.StringIO()
+
+    headers = set()
+    for datum in data:
+        if six.PY2:
```
Comment (amysutedja) [inline tests/searchcommands/chunked_data_stream.py:92]: Feels like we can remove this `if` statement and the one below.

File: tests/searchcommands/chunked_data_stream.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            headers.update(datum.keys())
+        else:
+            headers.update(datum.keys())
+    writer = csv.DictWriter(csvout, headers, dialect=splunklib.searchcommands.internals.CsvDialect)
+    writer.writeheader()
+    for datum in data:
+        if six.PY2:
+            writer.writerow(datum)
```
Comment (amysutedja) [inline tests/searchcommands/chunked_data_stream.py:100]: I'm probably missing something, but how do we know that the data values align with the order of the headers if we're making a set of potentially different-looking data?

**Review by amysutedja — CHANGES_REQUESTED** [review]

File: splunklib/searchcommands/search_command.py [hunk 8]
```diff
@@ -851,7 +851,8 @@ def _execute(self, ifile, process):
 
     @staticmethod
     def _as_binary_stream(ifile):
-        if six.PY2:
+        naught = ifile.read(0)
+        if isinstance(naught, bytes):
```
[the remark this answers is not in the document]
Comment (zenmoto) [reply splunklib/searchcommands/search_command.py:855]: Well, got it from you! :-)

File: tests/searchcommands/chunked_data_stream.py (hunk 1 above)
Comment (zenmoto) [reply tests/searchcommands/chunked_data_stream.py:50]: D'oh.

File: tests/searchcommands/chunked_data_stream.py (hunk 3 above)
Comment (zenmoto) [reply tests/searchcommands/chunked_data_stream.py:56]: In this case _build_data_csv explicitly returns a binary- in fact the last line of that function is `return six.ensure_binary(csvout.getvalue())`

File: tests/searchcommands/chunked_data_stream.py (hunk 4 above)
Comment (zenmoto) [reply tests/searchcommands/chunked_data_stream.py:61]: The idea here was to be able to parameterize this as needed to build custom searchinfo objects, but I didn't need anything more than these fields existing yet.

File: tests/searchcommands/chunked_data_stream.py (hunk 6 above)
Comment (zenmoto) [reply tests/searchcommands/chunked_data_stream.py:92]: Oh, yeah- I did get those unified. Thanks, I'll remove that.

File: tests/searchcommands/chunked_data_stream.py (hunk 7 above)
Comment (zenmoto) [reply tests/searchcommands/chunked_data_stream.py:100]: This is being managed by the `csv.DictWriter`. This looks like another place I had to take different paths for the pythons but then were able to re-converge them.

File: tests/searchcommands/chunked_data_stream.py (hunk 2 above)
Comment (zenmoto) [reply tests/searchcommands/chunked_data_stream.py:42]: Took me a second, but yes, you are right.

**Review by amysutedja — APPROVED** [review]
Approved pending passing tests.

**Pull request merged**